### PR TITLE
 📖 Add tutorial for external plugin

### PR DESCRIPTION
**Pull request by Eileen-Yu**
## Description:
Add docs for the external plugin tutorial.

## Motivation:
1. Guide users to write, install, and use their own external plugins.
2. Part of the efforts of https://github.com/kubernetes-sigs/kubebuilder/discussions/3309
3. Record my experiment & tips along with my practice over the external plugin with JavaScript:
https://github.com/Eileen-Yu/kb-js-plugin

### CODE REVIEW

File: docs/book/src/plugins/external-plugins.md
Comment (camilamacedo86) [inline docs/book/src/plugins/external-plugins.md:111]: it is missing edit

File: docs/book/src/plugins/external-plugins.md
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:101]:
```suggestion
- Your plugin is placed in a path following a `group-like name and version` scheme:
```

File: docs/book/src/plugins/external-plugins.md
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:13]: Should we still use the terminology as "internal plugins" since all plugins will be implemented as external plugins?

File: docs/book/src/plugins/external-plugins.md
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:7]: Could you elaborate a bit more on what you mean by same kubebuilder plugin features? It might be nice to include those here.

File: docs/book/src/plugins/external-plugins.md
Comment (rashmigottipati) [reply docs/book/src/plugins/external-plugins.md:103]: I like @camilamacedo86's suggestion.
> WDYT about something more like ^ ?

```suggestion
- kubebuilder CLI > 3.11.0
- Your own external plugin executable written in a language of your choice 
- Path of the external plugin configured via the ENV VAR (specify the environment variable here) or it can be placed in a path following a `group-like name and version` scheme:

```

Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:103]:
```suggestion
 $HOME/.config/kubebuilder/plugins/${name}/${version} for Linux
 ~/Library/Application Support/kubebuilder/plugins/${name}/${version} for OSX
```

File: docs/book/src/plugins/external-plugins.md
Comment (rashmigottipati) [inline docs/book/src/plugins/external-plugins.md:133]: Instead of `/path1/path2` can we specify something else as it may cause confusion. For example:
```suggestion
EXTERNAL_PLUGINS_PATH=<home-dir>/sampleplugin/v1/sampleplugin kubebuilder create api --plugins sampleplugin/v1 --number 2
```

This comment applies to all other places where `/path1/path2` is mentioned.

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:5]: Nit: I would reword this a bit to be more along the lines of:
```suggestion
Kubebuilder's functionality can be extended through the use of external plugins.
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:7]: +1 to more elaboration. I would also recommend a bit of a rephrase here to be something more along the lines of:
```suggestion
An external plugin is an executable that implements an execution pattern that Kubebuilder knows how to interact with. Since an external plugin is an executable it can be implemented in any language. 
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:13]: +1 on avoiding "internal plugins" as terminology here. I would probably say something more along the lines of "Kubebuilder's default scaffolding"

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:17]:
```suggestion
- If you are looking for implementing plugins in a language other than `Go`.
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:79]: It might be nice to include an example of the JSON for both of these. It might be difficult for someone reading the docs to visualize the JSON based on Go type definitions.

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [reply docs/book/src/plugins/external-plugins.md:111]: We should also add the optional subcommands for the plugin that provides metadata to Kubebuilder

**Review by everettraven — COMMENTED** [review]
Great start @Eileen-Yu ! Here are some comments from my first review pass:

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:126]: Thanks for catching! 
And based on @rashmigottipati 's suggestion, hope using `export EXTERNAL_PLUGINS_PATH=<custom-path>` could be more clear 🤔

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:9]: Nit: Some suggestions for wording improvement:

```suggestion
The Kubebuilder CLI loads the external plugin in the specified path and interacts with it through `stdin` & `stdout`.
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:23]: Nit: "interfaces" suggests a looser construct in programming ("interfaces" can be anything behind the scenes as long as they implement certain functionality). I would recommend using "types" as it is more strict and aligns with what the external plugin communication process in Kubebuilder expects. The response must be exactly the `PluginResponse` type and shouldn't have any additional fields

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:97]: I'm not sure this should be a prerequisite - I'm interpreting this as I _have_ to write my own external plugin rather than being able to use any external plugin

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:123]: nit:
```suggestion
If a plugin does not implement the `flags` subcommand, Kubebuilder will pass all flags to the plugin, making it the external plugin's responsibility to handle any invalid flags. 
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:122]: Nit:
```suggestion
The `flags` subcommand in an external plugin allows for early error detection by informing Kubebuilder about the flags the plugin supports. If an unsupported flag is identified, Kubebuilder can issue an error before the plugin is called to execute. 
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:125]: nit:
```suggestion
If a plugin does implement the `flags` subcommand, it **MUST** support `--init`, `--api`, `--webhook`, and `--edit` flags, which correspond to respective Kubebuilder subcommands.
```

**Review by everettraven — COMMENTED** [review]
This doc is starting to look really good! A few more nits on my second pass:

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Good idea!
I added sample JSONs for both request & response together with explainations.
Also, the Go definitions are now referred in the link instead of directly pasting them in the docs.

Hope this can make it clearer..

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Thank you for the nice explanation! 👍
I tried to reorganized this paragraph in the latest commit.

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md]: Oh I see!
So by this line I tried to say the external plugin needs to be an executable..
I updated this part with some explanations.

File: docs/book/src/plugins/external-plugins.md
Comment (Kavinjsir) [inline docs/book/src/plugins/external-plugins.md:28]:
```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by "kubebuilder init --plugins v1alpha1 --domain my.domain:
```

File: docs/book/src/plugins/external-plugins.md
Comment (Kavinjsir) [inline docs/book/src/plugins/external-plugins.md:7]:
```suggestion
An external plugin is an executable (can be implemented in any language) that implements an execution pattern that Kubebuilder knows how to interact with.
```

**Review by Kavinjsir — APPROVED** [review]
LGTM!
Left some nits which can or cannot be applied.

**Review by Kavinjsir — APPROVED** [review]

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:47]:
```suggestion
    	 "examples": "kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain"
```

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:28]: Thanks for the nice suggestions👍
Just applied!

**Review by rashmigottipati — APPROVED** [review]
Nice work @Eileen-Yu, changes look good to me!

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:28]: nit:
```suggestion
Below is a sample JSON object of the `PluginRequest` type, triggered by `kubebuilder init --plugins sampleexternalplugin/v1 --domain my.domain`:
```

File: docs/book/src/plugins/external-plugins.md
Comment (everettraven) [inline docs/book/src/plugins/external-plugins.md:111]: Nit: Now that I am thinking about it a bit more I don't think these flags are _required_ to be implemented as the external plugin would return an error about an unknown flag that Kubebuilder would ignore. That being said, we may want to consider the pros and cons of enforcing or not enforcing this behavior. 

@Eileen-Yu I'll leave it up to you if you want to change the wording to reflect that these flags aren't _required_ . IMO though this nit shouldn't block this PR from merging.

**Review by everettraven — APPROVED** [review]
This looks great @Eileen-Yu ! Thanks for putting this much needed documentation together!

I do have a couple teeny non-blocking nits:

File: docs/book/src/plugins/external-plugins.md
Comment (Eileen-Yu) [reply docs/book/src/plugins/external-plugins.md:111]: I can understand that flag for `--init` … is more like a ‘bonus’ instead of ‘must’. For me, actually I’m still curious about the detailed scenario to support that so as to reach “fast fail”.

So I guess maybe we can drop that paragraph includes the 'must' literal.
I was imagining if it is fair enough to stay focus on the general usage(`init`, `create` and `edit`) of the external plugin.

**Comment by everettraven** [comment]
@camilamacedo86 could you give this an approval whenever you have a chance?

**Review by camilamacedo86 — APPROVED** [review]
/lgtm

**Pull request merged**
